switch requirements from depcheck to rpmdeplint, rework resultsdb queries

kparal opened the pull request:
Depcheck is replaced by rpmdeplint, update the code. Also fix queries to
resultsdb:
* use API 2.0
* ask for both bodhi_update and koji_build results
* do very specific queries (only request required testcases, use
timestamps)
* use results/latest endpoint by default
* give up if there are too many results for a given item (doesn't affect
results/latest endpoint, only results endpoint)
* use larger pages when retrieving testcases

There are known issues with the current results/latest endpoint
approach, in particular it doesn't display results for all architectures
(for rpmdeplint). Since rpmdeplint is currently only executed for
x86_64, it isn't a huge problem atm, but will need to be resolved in the
future.

----

Hello, this is a preliminary patch for testing and discussion. Rpmdeplint is only run in dev at the moment, we'll be pushing it into production after Beta freeze (so this patch doesn't make sense before that). Furthermore, this patch is completely untested. I used a development VM using vagrant, but I have no idea how to actually test this backend stuff. (I probably broke the unit tests as well). Can you please test it?

Is the code in `check_requirements()` even used at all? And when it is used? How can I trigger it when using vagrant dev env?

I assume that `check_requirements()` should block karma autopush in case any of those testcases specified in `site_requirements` are not passed. Is that correct? Is it being used at the moment? (Because it seems to me that this must've been broken for some time).

Thanks.

kparal committed:
switch requirements from depcheck to rpmdeplint, rework resultsdb queries

Depcheck is replaced by rpmdeplint, update the code. Also fix queries to
resultsdb:
* use API 2.0
* ask for both bodhi_update and koji_build results
* do very specific queries (only request required testcases, use
timestamps)
* use results/latest endpoint by default
* give up if there are too many results for a given item (doesn't affect
results/latest endpoint, only results endpoint)
* use larger pages when retrieving testcases
* don't filter ABORTED results, they can no longer be stored in
resultsdb

There are known issues with the current results/latest endpoint
approach, in particular it doesn't display results for all architectures
(for rpmdeplint). Since rpmdeplint is currently only executed for
x86_64, it isn't a huge problem atm, but will need to be resolved in the
future.

bowlofeggs commented:
Hey @kparal!

This pull request does cause quite a few test failures and errors (13 and 3, respectively), as you guessed ☺ When you are in the vagrant box you can run the tests with ```btest```. You can also use handy flags, like ```-x``` to exit upon the first failure. When you ```vagrant ssh``` into the box, you should see a handy motd printed to your terminal that tells you about btest and some other handy aliases.

Here's a snippet of one of the errors:

```
======================================================================
FAIL: Test submitting a stable request, but with bad taskotron results
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/home/vagrant/bodhi/bodhi/tests/server/services/test_updates.py", line 2330, in test_stable_request_failed_taskotron_results
    self.assertIn('Required task', resp)
AssertionError: 'Required task' not found in <400 Bad Request application/json body='{"status"..."}]}'/176>
-------------------- >> begin captured logging << --------------------
bodhi: INFO: Using the FakeBugTracker
bodhi.server: DEBUG: Creating all models for Engine(sqlite://)
bodhi.server: DEBUG: Setting alias for bodhi-2.0-1.fc17 to FEDORA-2017-a3bbe1a8f2
bodhi: INFO: Using the FakeBugTracker
dogpile.lock: DEBUG: NeedRegenerationException
dogpile.lock: DEBUG: no value, waiting for create lock
dogpile.lock: DEBUG: value creation lock <dogpile.cache.region._LockWrapper object at 0x7f50cc581290> acquired
dogpile.lock: DEBUG: Calling creation function
dogpile.lock: DEBUG: Released creation lock
bodhi.server: DEBUG: validated = {'close_bugs': True, 'stable_karma': 3, 'csrf_token': u'f6a9de0b52ec76952a0400e69586b0421fed9952', 'requirements': u'rpmlint', 'builds': [u'bodhi-2.0.0-3.fc17'], 'autokarma': True, 'edited': '', 'suggest'
: <unspecified>, 'notes': u'this is a test update', 'request': <testing>, 'bugs': [], 'unstable_karma': -3, 'require_bugs': False, 'require_testcases': True, 'type': <bugfix>, 'severity': <unspecified>}
bodhi.server: DEBUG: Adding nvr bodhi-2.0.0-3.fc17
bodhi.server: INFO: Creating new update: [<RpmBuild {'epoch': 0, 'type': u'rpm', 'nvr': u'bodhi-2.0.0-3.fc17', 'signed': False}>]
bodhi.server: DEBUG: Creating new Update(**data) object.
bodhi.server: DEBUG: Assigning alias for new update..
bodhi.server: DEBUG: Setting alias for bodhi-2.0.0-3.fc17 to FEDORA-2017-48150de3bc
bodhi.server: DEBUG: Setting request for new update.
bodhi.server: DEBUG: Attempting to set request <testing>
bodhi.server: DEBUG: Adding tag f17-updates-testing-signing to bodhi-2.0.0-3.fc17
bodhi: DEBUG: tagBuild(f17-updates-testing-signing, bodhi-2.0.0-3.fc17)
bodhi.server: DEBUG: bodhi-2.0.0-3.fc17 has been submitted for testing. 
bodhi.server: INFO: Sending mail to guest: [Fedora Update] [comment] bodhi-2.0.0-3.fc17
bodhi.server: INFO: Not sending email: No smtp_server defined
bodhi.server: DEBUG: Adding new update to the db.
bodhi.server: DEBUG: Triggering db flush for new update.
bodhi.server: DEBUG: Done with Update.new(...)
bodhi.server: DEBUG: bodhi-2.0.0-3.fc17 update created
bodhi.server: INFO: Sending mail to guest: [Fedora Update] [comment] bodhi-2.0.0-3.fc17
bodhi.server: INFO: Not sending email: No smtp_server defined
dogpile.lock: DEBUG: NeedRegenerationException
dogpile.lock: DEBUG: no value, waiting for create lock
dogpile.lock: DEBUG: value creation lock <dogpile.cache.region._LockWrapper object at 0x7f50ccbf1890> acquired
dogpile.lock: DEBUG: Calling creation function
dogpile.lock: DEBUG: Released creation lock
bodhi.server: DEBUG: validating request
--------------------- >> end captured logging << ---------------------
```

As for ```check_requirements()```, I do see calls to it in two places:

0. [The masher seems to call it](https://github.com/fedora-infra/bodhi/blob/2.6.2/bodhi/server/consumers/masher.py#L434-L437) before sending the update out, and seems to reject it if result doesn't evaluate "truthy".
1. The [set_request update handler seems to call it](https://github.com/fedora-infra/bodhi/blob/2.6.2/bodhi/server/services/updates.py#L127-L133) when the update is requested for stable via the API.

Weirdly, it doesn't seem to get used when an update gets set for stable based on karma, though the masher theoretically would still reject it.

So to trigger it in the vagrant environment, I'd say make a call with the CLI (which is automatically configured to talk to localhost in the vagrant box) to request stable. In order to do that, you will also need to fake the update meeting testing requirements. You can make it look like an update has been in testing for ```n``` days with bshell:

```
$ bshell
<snip>
In [1]: u = m.Update.query.filter(m.Update.alias=="FEDORA-2017-afd3f53b9d").one()
/usr/lib64/python2.7/site-packages/sqlalchemy/sql/sqltypes.py:185: SAWarning: Unicode type received non-unicode bind param value 'FEDORA-2017-afd3f53b9d'. (this warning may be suppressed after 10 occurrences)
  (util.ellipses_string(value),))

In [3]: import datetime

In [4]: u.date_testing = datetime.datetime.now() - datetime.timedelta(days=15)

In [5]: s().add(u)

In [6]: s().commit()
```

That will make the update be there long enough, and then you can just do the "normal" things to try to push it to stable and that should call ```check_requirements()```.

Let me know if you need further help getting the tests to pass, or if you think I need to explain something more clearly!

bowlofeggs commented on bodhi/server/models.py:
```diff
@@ -2056,24 +2056,48 @@ def check_requirements(self, session, settings):
+            buildinfos = koji.multiCall()
+
+            for index, build in enumerate(self.builds):
+                buildinfo = buildinfos[index]
```
I'm not very familiar with the Koji API - are the results of the multCall() guaranteed to be in the same order as the nvrs we passed above? Thus making this buildinfo definitely correlate to the build?

kparal replied:
Yes, this is the docstring:

>     def multiCall(self, strict=False):
>         """Execute a multicall (multiple function calls passed to the server
>         and executed at the same time, with results being returned in a batch).
>         Before calling this method, the self.multicall field must have
>         been set to True, and then one or more methods must have been called on
>         the current session (those method calls will return None).  On executing
>         the multicall, the self.multicall field will be reset to False
>         (so subsequent method calls will be executed immediately)
>         and results will be returned in a list.  The list will contain one element
>         for each method added to the multicall, in the order it was added to the multicall.
>         Each element of the list will be either a one-element list containing the result of the
>         method call, or a map containing "faultCode" and "faultString" keys, describing the
>         error that occurred during the method call."""
>

bowlofeggs commented on bodhi/server/models.py:
```diff
@@ -2056,24 +2056,48 @@ def check_requirements(self, session, settings):
+                buildinfo = buildinfos[index]
+                if not isinstance(buildinfo, dict):
+                    log.warn("Koji build info doesn't seem correct: %r",
+                             buildinfo)
```
It seems like we'd want to raise an Exception here, perhaps a ```TypeError```.

kparal replied:
It will fail on the next step, probably raising `TypeError`. Do you think I should raise that immediately?

bowlofeggs commented on bodhi/server/models.py:
```diff
@@ -2056,24 +2056,48 @@ def check_requirements(self, session, settings):
+                results.extend(build_results)
+
+        except Exception as e:
+            return False, "Failed retrieving requirements results: %r" % e.message
```
We should probably do a log.exception() here too so we can have a record of what happened.

kparal replied:
Good idea, will do.

bowlofeggs commented on bodhi/server/models.py:
```diff
@@ -2056,24 +2056,48 @@ def check_requirements(self, session, settings):
 
             for arch, results in by_arch.items():
-                latest = results[0]  # TODO - do these need to be sorted still?
+                latest = results[0]  # resultsdb results are ordered chronologically
```
Thanks for answering the question ☺

bowlofeggs commented on bodhi/server/util.py:
```diff
@@ -594,27 +594,44 @@ def tokenize(string):
+        in the database and it's very reasonable to limit queries (thus the
+        default value).
+    :param kwargs: Any args that will be passed to resultsdb to specify what
+        results to retrieve.
```
+1 I love documented parameters!

bowlofeggs commented on bodhi/server/util.py:
```diff
@@ -594,27 +594,44 @@ def tokenize(string):
+            url = json.get('next')
+            queries += 1
+            if queries >= max_queries and url:
+                log.warn('Too many result pages, aborting at: %r' % url)
```
Should this be a warning, or maybe just an info or even a debug? After all, the caller asked for ```max_queries``` so I'm not sure we should warn that we hit it.

kparal replied:
You're right, I wrote this before I added `max_queries` as a method argument. Now it's much better to have it as debug.

bowlofeggs commented on production.ini:
```diff
@@ -204,9 +204,7 @@ stacks_enabled = False
-site_requirements = depcheck upgradepath
-## Some day we'll have rpmgrill, and that will be cool.  Ask tflink.
-#site_requirements = depcheck upgradepath rpmgrill
+site_requirements = dist.rpmdeplint dist.upgradepath
```
I've got a PR open to rework basically everything about how settings in Bodhi work: #1544. If that gets merged first, we should make sure we make these changes in the new ```bodhi.server.config.BodhiConfig._defaults```. If not, I'll have to be sure to add this there.

bowlofeggs replied:
#1544 is now merged.

bowlofeggs reviewed (changes requested):
I've got a few notes for what you have so far too. Thanks for the PR!

kparal commented:
> When you are in the vagrant box you can run the tests with btest. You can also use handy flags, like -x to exit upon the first failure.

I was hoping you do the tests for me :-P OK, will try.

> As for check_requirements(), I do see calls to it in two places:
> 
>     The masher seems to call it before sending the update out, and seems to reject it if result doesn't evaluate "truthy".
>     The set_request update handler seems to call it when the update is requested for stable via the API.
> 
> Weirdly, it doesn't seem to get used when an update gets set for stable based on karma, though the masher theoretically would still reject it.

When the masher rejects it, is it somehow visible in Bodhi UI (a comment added to the update)? I remember seeing a comment like "autokarma push disabled, because an important test failed" (or something along those lines). Is that it, or is it something else? And where does that code live?

There's one important piece of information. Due to how our tests are currently written, `depcheck` is executed immediately for everything in `fxx-updates-testing-pending` and `fxx-updates-pending`, but `upgradepath` is only executed for packages in `fxx-updates-pending`. So until the user manually requests a push to stable (or autokarma limit is reached), the package has no `upgradepath` result. As I currently see the code, that would prevent manual pushing to updates. And it could also make a race condition with the masher, if the result doesn't arrive in time before the masher processes it. But that doesn't seem to actually happen (updates can be pushed stable manually), so either the code is not really being used, or I'm missing something. Thoughts?

> You can make it look like an update has been in testing for n days with bshell:

Is it possible to give myself superuser permissions to do anything and everything, without tinkering with the database? I added a group I'm in into `admin_packager_groups` in `development.ini`, and that allowed me to create updates for any package, but I still can't push it until the requirements are met (days in testing, karma -- and I can't give karma to my updates). Thanks for that bshell help, though!

kparal committed:
fix tests for the previous commit

Also fix a few bugs.

kparal committed:
improve performance of requirement validation

When checking whether a taskotron testcase exists, don't pull all
existing testcases, but rather ask for just that particular testcase.

kparal commented:
It was a tough fight, but the test suite is working now :) I had to do some major changes, please review. I also added more optimizations.

Bodhi UI is working for me, I can set required testcases (name verification works), I can't push stable until they pass (nice). I don't know how to verify the masher, but the test suite runs...

What I think doesn't work and hasn't worked are `site_requirements`. It seems they are really not used anywhere, or at least I don't see it. If I create a new update, by default the testcases field is empty, and it means "no requirements". So the site defaults doesn't really seem to be used. I don't know if that's a problem or not, as I said, we still haven't fine-tuned all the processes (especially the upgradepath is getting run late), so it might be fine having it just like this for now.

What I'd like to see and verify is that karma autopush is disabled if any of `site_requirements` fail. I think Bodhi should have such a feature already, but I can't locate it.

bowlofeggs commented on bodhi/server/buildsys.py:
```diff
@@ -25,7 +26,23 @@
+    return wrapper
+
+
+class Buildsystem(object):
```
Is this change needed for the goal at hand? If not, I suggest making it a separate pull request.

bowlofeggs replied:
By "this change" I mean explicitly subclassing object here.

kparal replied:
It's needed for a class to have properties. I made `multicall` a property, in order to fix previous incorrect multicall implementation (it wasn't possible to test multicall on methods which called each other).

bowlofeggs commented on bodhi/server/models.py:
```diff
@@ -2049,31 +2049,66 @@ def check_requirements(self, session, settings):
         requirements = list(requirements)
 
+        if not requirements:
+            return True, "All checks pass."
```
Maybe change "All checks pass." to "No checks required."

bowlofeggs commented on bodhi/server/templates/update.html:
```diff
@@ -128,10 +128,6 @@
 
-        // However, we want to skip over any ABORTED results:
-        // https://github.com/fedora-infra/bodhi/issues/167
-        if (result.outcome == 'ABORTED') return;
```
Do we not want to skip aborted results anymore?

kparal replied:
We no longer allow resultsdb to store them.

bowlofeggs commented on development.ini.example:
```diff
@@ -216,9 +216,7 @@ stacks_enabled = False
-site_requirements = depcheck upgradepath
-## Some day we'll have rpmgrill, and that will be cool.  Ask tflink.
-#site_requirements = depcheck upgradepath rpmgrill
+site_requirements = dist.rpmdeplint dist.upgradepath
```
Please rebase this PR on the develop branch - once you've done that, please also make this setting commented by default (similar to most of the rest of the file) and then add this default to ```bodhi.server.config.BodhiConfig._defaults```. Thanks!

bowlofeggs reviewed (changes requested):
The tests pass, but there are a number of lines of code that are missing test coverage:

```
$ diff-cover coverage.xml --compare-branch=develop --fail-under=100
Failure. Coverage is below 100.0%.
-------------
Diff Coverage
Diff: develop...HEAD, staged, and unstaged changes
-------------
bodhi/server/buildsys.py (100%)
bodhi/server/models.py (75.9%): Missing lines 2059,2081,2083-2084,2094-2096
bodhi/server/util.py (9.1%): Missing lines 609-610,614,617,623,626-630
bodhi/server/validators.py (42.9%): Missing lines 929-930,932-933
bodhi/server/views/generic.py (100%)
-------------
Total:   75 lines
Missing: 21 lines
Coverage: 72%
```

kparal committed:
address review requests, add more tests, fix bugs

kparal commented:
After updating `develop`, I can't reprovision vagrant anymore, even if I do `vagrant destroy && vagrant up`. Seems something broke:
```
TASK [dev : Apply database migrations] *****************************************
fatal: [bodhi]: FAILED! => {"changed": true, "cmd": ["alembic", "upgrade", "head"], "delta": "0:00:00.920123", "end": "2017-06-07 13:08:01.694655", "failed": true, "rc": 1, "start": "2017-06-07 13:08:00.774532", "stderr": "Traceback (most recent call last):\n  File \"/bin/alembic\", line 9, in <module>\n    load_entry_point('alembic==0.8.6', 'console_scripts', 'alembic')()\n  File \"/usr/lib/python2.7/site-packages/alembic/config.py\", line 479, in main\n    CommandLine(prog=prog).main(argv=argv)\n  File \"/usr/lib/python2.7/site-packages/alembic/config.py\", line 473, in main\n    self.run_cmd(cfg, options)\n  File \"/usr/lib/python2.7/site-packages/alembic/config.py\", line 456, in run_cmd\n    **dict((k, getattr(options, k)) for k in kwarg)\n  File \"/usr/lib/python2.7/site-packages/alembic/command.py\", line 174, in upgrade\n    script.run_env()\n  File \"/usr/lib/python2.7/site-packages/alembic/script/base.py\", line 397, in run_env\n    util.load_python_file(self.dir, 'env.py')\n  File \"/usr/lib/python2.7/site-packages/alembic/util/pyfiles.py\", line 93, in load_python_file\n    module = load_module_py(module_id, path)\n  File \"/usr/lib/python2.7/site-packages/alembic/util/compat.py\", line 79, in load_module_py\n    mod = imp.load_source(module_id, path, fp)\n  File \"alembic/env.py\", line 17, in <module>\n    config.set_main_option(\"sqlalchemy.url\", bodhi_config[\"sqlalchemy.url\"])\n  File \"/home/vagrant/bodhi/bodhi/server/config.py\", line 551, in __getitem__\n    self.load_config()\n  File \"/home/vagrant/bodhi/bodhi/server/config.py\", line 584, in load_config\n    self._validate()\n  File \"/home/vagrant/bodhi/bodhi/server/config.py\", line 602, in _validate\n    'Invalid config values were set: \\n{}'.format('\\n'.join(errors)))\nValueError: Invalid config values were set: \n\tcompose_atomic_trees: \"\" cannot be interpreted as a boolean value.", "stderr_lines": ["Traceback (most recent call last):", "  File \"/bin/alembic\", line 9, in <module>", "    load_entry_point('alembic==0.8.6', 'console_scripts', 'alembic')()", "  File \"/usr/lib/python2.7/site-packages/alembic/config.py\", line 479, in main", "    CommandLine(prog=prog).main(argv=argv)", "  File \"/usr/lib/python2.7/site-packages/alembic/config.py\", line 473, in main", "    self.run_cmd(cfg, options)", "  File \"/usr/lib/python2.7/site-packages/alembic/config.py\", line 456, in run_cmd", "    **dict((k, getattr(options, k)) for k in kwarg)", "  File \"/usr/lib/python2.7/site-packages/alembic/command.py\", line 174, in upgrade", "    script.run_env()", "  File \"/usr/lib/python2.7/site-packages/alembic/script/base.py\", line 397, in run_env", "    util.load_python_file(self.dir, 'env.py')", "  File \"/usr/lib/python2.7/site-packages/alembic/util/pyfiles.py\", line 93, in load_python_file", "    module = load_module_py(module_id, path)", "  File \"/usr/lib/python2.7/site-packages/alembic/util/compat.py\", line 79, in load_module_py", "    mod = imp.load_source(module_id, path, fp)", "  File \"alembic/env.py\", line 17, in <module>", "    config.set_main_option(\"sqlalchemy.url\", bodhi_config[\"sqlalchemy.url\"])", "  File \"/home/vagrant/bodhi/bodhi/server/config.py\", line 551, in __getitem__", "    self.load_config()", "  File \"/home/vagrant/bodhi/bodhi/server/config.py\", line 584, in load_config", "    self._validate()", "  File \"/home/vagrant/bodhi/bodhi/server/config.py\", line 602, in _validate", "    'Invalid config values were set: \\n{}'.format('\\n'.join(errors)))", "ValueError: Invalid config values were set: ", "\tcompose_atomic_trees: \"\" cannot be interpreted as a boolean value."], "stdout": "", "stdout_lines": []}
	to retry, use: --limit @/home/kparal/devel/forks/bodhi/devel/ansible/playbook.retry
```

bowlofeggs commented:
On Wed, 2017-06-07 at 05:20 -0700, Kamil Páral wrote:
> It's needed for a class to have properties. I made multicall a
> property, in order to fix previous incorrect multicall implementation
> (it wasn't possible to test multicall on methods which called each
> other).

Cool.

bowlofeggs commented:
On Wed, 2017-06-07 at 05:22 -0700, Kamil Páral wrote:
> We no longer allow resultsdb to store them.

OK, thanks!

kparal commented:
OK, tests added. I hope this is the last iteration, because I can't really spend more time on this.

bowlofeggs reviewed (approved):
Looks great - very nice tests. I'm going to squash the commits down to one and merge. Thanks!

kparal commented:
Thanks @bowlofeggs .
